feat(codegen): have `with_source_text` reserve memory for code buffer

DonIsaac opened the pull request:
`Codegen::with_source_text` gives us enough information to use `Codegen::with_capacity`. This change makes the API cleaner as users have to provide less redundant information.

overlookmotel commented on crates/oxc_codegen/src/lib.rs:
```diff
@@ -158,7 +167,9 @@ impl<'a> Codegen<'a> {
     #[must_use]
     pub fn with_capacity(mut self, source_text_len: usize) -> Self {
         let capacity = if self.options.minify { source_text_len / 2 } else { source_text_len };
-        self.code = Vec::with_capacity(capacity);
+        // ensure space for at least `capacity` additional bytes without clobbering existing
+        // allocations.
+        self.code.reserve(capacity);
```
I must say, I don't understand `Codegen`'s API. I'd have expected it'd be impossible to add to `self.code` - I'd have expected you just call `build`, and that consumes the `Codegen`.

But no! We have public `print_char`, `print_str` and `print_expression` methods which push to `self.code`. I'm not sure if that's a mistake or not.

But as the API stands now, it looks like `self.code.reserve(capacity)` is correct, so as not to clobber any existing data in `self.code`.

DonIsaac replied:
I agree, its API is quite confusing. For example, there are 3 methods that take source_text. 

The linter uses Codegen in fixers, where providing source_text and an entire AST is not desireable. Maybe we should separate out expression printing and provide two different APIs for those two different use cases?

overlookmotel replied:
> Maybe we should separate out expression printing and provide two different APIs for those two different use cases?

I would say that's a good idea, yes.

DonIsaac committed:
feat(codegen): have `with_source_text` reserve memory for code buffer (#5823)

`Codegen::with_source_text` gives us enough information to use `Codegen::with_capacity`. This change makes the API cleaner as users have to provide less redundant information.
